Document/0x05h-Testing-Platform-Interaction.md

```diff
@@ -1176,9 +1176,96 @@ There are several ways to perform dynamic analysis:
+                  // Request the update.
+
+
+                  //...Part 2: request update
+                  appUpdateManager.startUpdateFlowForResult(
+                     // Pass the intent that is returned by 'getAppUpdateInfo()'.
+                     appUpdateInfo,
+                     // Or 'AppUpdateType.FLEXIBLE' for flexible updates.
+                     AppUpdateType.IMMEDIATE,
+                     // The current activity making the update request.
+                     this,
+                     // Include a request code to later monitor this update request.
+                     MY_REQUEST_CODE);
+
+
+                     //...Part 3: check if update completed succesfully
+ @Override
+ public void onActivityResult(int requestCode, int resultCode, Intent data) {
+   if (myRequestCode == MY_REQUEST_CODE) {
+     if (myRequestCode != RESULT_OK) {
+       log("Update flow failed! Result code: " + resultCode);
+       // If the update is cancelled or fails,
+       // you can request to start the update again in case of forced updates
+     }
+   }
+ }
+
+ //..Part 4:
+ // Checks that the update is not stalled during 'onResume()'.
+// However, you should execute this check at all entry points into the app.
+@Override
+protected void onResume() {
+  super.onResume();
+
+  appUpdateManager
+      .getAppUpdateInfo()
+      .addOnSuccessListener(
+          appUpdateInfo -> {
+            ...
+            if (appUpdateInfo.updateAvailability()
+                == UpdateAvailability.DEVELOPER_TRIGGERED_UPDATE_IN_PROGRESS) {
+                // If an in-app update is already running, resume the update.
+                manager.startUpdateFlowForResult(
+                    appUpdateInfo,
+                    IMMEDIATE,
+                    this,
+                    MY_REQUEST_CODE);
+            }
+          });
+}
+}
+```
+>Source: https://developer.android.com/guide/app-bundle/in-app-updates
+
+When checking for a proper update mechanism, make sure the usage of the `AppUpdateManager` is present. If it is not yet, then this means that users might be able to remain on an older version of the application with the given vulnerabilities.
+Next, pay attention to the `AppUpdateType.IMMEDIATE` use: if a security update comes in, then this flag should be used in order to make sure that the user cannot go forward with using the app without updating it.
+Last, in part 3 of the example: make sure that cancelations or errors do end up in re-checks and that a user cannot move forward in case of a critical security update.
+
+#### Dynamic analysis
+In order to test for proper updating: try downloading an older version of the application with a security vulnerability, either by a release from the developers or by using a third party app-store. Next, see if it requires the enforced updating and if you are able to cancel the update one way or another and if you can then still continue to use the app / exploit the vulnerability.
```

Comment:
Bypassing this check will always be possible, through hooking, rewriting, etc. I'm guessing you mean "check if a normal user can bypass it, for example by pressing cancel" ?

Reply:
indeed, or a scriptkiddy by doing odd stuff